csrc/ops/composite.cpp

```diff
@@ -89,6 +89,40 @@ TensorView* matmul(TensorView* a, TensorView* b) {
   return matmul(a, b, true /* cast output to input dtype */);
 }
 
+TensorView* linear(TensorView* a, TensorView* b, TensorView* bias) {
+  NVF_CHECK(
+      a->nDims() == b->nDims(),
+      "The number of dimension of A and B do not match");
+  NVF_CHECK(a->nDims() == 2, "Only 2-D A/B Tensors are supported in Linear!");
+
+  std::vector<bool> bcast_dims(a->nDims() + 1, false);
+  // A: [M, Bcast, K]
+  // B: [Bcast, N, K]
+  bcast_dims.at(bcast_dims.size() - 2) = true;
+  auto* tv0b = broadcast(a, bcast_dims);
+  bcast_dims.at(bcast_dims.size() - 2) = false;
+  bcast_dims.at(bcast_dims.size() - 3) = true;
+  auto* tv1b = broadcast(b, bcast_dims);
+
+  NVF_CHECK(
+      a->getDataType().value() == b->getDataType().value(),
+      "data types of inputs to matmul don't match");
+
+  auto* output = fusedMultiplySum(tv0b, tv1b, {-1});
+  if (bias) {
+    NVF_CHECK(bias->nDims() == 1, "Only 1-D Bias are supported in Linear!");
```

Comment:
We should also check that the shape of bias is `N,`

Reply:
Also, that the bias is of the same type as the inputs.

Reply:
> We should also check that the shape of bias is N,
What would be the way to do this for symbolic tensors?

Reply:
can we actually check the shape of a tensor?

Reply:
nitpick: is `bias->nDims() == 1` a hard requirement? shouldn't we be able to just do normal broadcasting on `bias`?

I know torch doc suggests that bias should be shape `(out_features)` https://pytorch.org/docs/stable/generated/torch.nn.functional.linear.html#torch.nn.functional.linear

But implementation seems to disagree: 
```
>>> a = torch.randn(8, 2, 3, 4, device="cuda")
>>> b = torch.randn(5, 4, device="cuda")
>>> bias = torch.randn(8, 2, 3, 5, device="cuda")
>>> torch.ops.aten.linear(a, b, bias).shape
torch.Size([8, 2, 3, 5])
>>> torch.nn.functional.linear(a, b, bias).shape
torch.Size([8, 2, 3, 5])
```

Reply:
> We should also check that the shape of bias is `N,`

We can't check that at definition. It will be assumed to be true then when we receive inputs, if it is untrue we will display a somewhat cryptic error message in `ExpressionEvaluator::propagateBoundValuesThroughExactMaps`. :face_with_spiral_eyes:

Reply:
> But implementation seems to disagree:

Interesting. So this should just be "torch-broadcastable" to match the output, i.e. bias can be any dimension up to `a->nDims()` and we should do something like `ops::maybeBroadcast({a, bias})`:
https://github.com/NVIDIA/Fuser/blob/356a59d7f9b99238cd6c330b00c883903fbee8f8/csrc/ops/utils.cpp#L296. But whereas `maybeBroadcast` inserts the broadcasts as new inner dimensions, we should instead add broadcasts in the outer dims. BTW @jjsjann123 do you happen to know why our implicit broadcasting works differently from pytorch?
```python
[ins] In [1]: import torch
[nav] In [2]: x = torch.randn([4, 5])
[nav] In [3]: y = torch.randn([5])
[ins] In [4]: x + y
Out[4]: 
tensor([[-1.8200,  2.8528, -1.2850, -2.3024, -0.1085],
        [-2.7299, -0.3150,  0.2120,  1.2288, -0.4108],
        [-0.8917, -0.1539, -0.3934,  0.6995, -0.1939],
        [-1.7878,  0.4470, -2.2318,  0.1037,  0.4602]])
```

Reply:
Oh, I think I misread this code! It is actually broadcasting the outer dims.

Reply:
forwarding Kevin's comment, looks like we are just matching thunder's checks: https://github.com/NVIDIA/Fuser/issues/2057#issuecomment-2064838196